dm/ctl/master/operate_task.go

```diff
@@ -0,0 +1,194 @@
+// limitations under the License.
+
+package master
+
+import (
+	"context"
+	"errors"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/spf13/cobra"
+
+	"github.com/pingcap/dm/dm/ctl/common"
+	"github.com/pingcap/dm/dm/pb"
+)
+
+const (
+	batchSizeFlag    = "batch-size"
+	defaultBatchSize = 5
+)
+
+type batchTaskResult struct {
+	Result       bool                 `json:"result"`
+	Msg          string               `json:"msg"`
+	BatchResults []*operateTaskResult `json:"batchResults"`
+}
+
+type operateTaskResult struct {
+	TaskName string                     `json:"taskName"`
+	Op       string                     `json:"op"`
+	Result   bool                       `json:"result"`
+	Msg      string                     `json:"msg"`
+	Sources  []*pb.CommonWorkerResponse `json:"sources"`
+}
+
+func operateTaskFunc(taskOp pb.TaskOp, cmd *cobra.Command) error {
+	allMode, err := cmd.Flags().GetBool("all")
+	if err != nil {
+		common.PrintLinesf("error in parse `--all`")
+		return err
+	}
+	if allMode {
+		return operateSourceTaskFunc(taskOp, cmd)
+	}
+
+	if len(cmd.Flags().Args()) != 1 {
+		cmd.SetOut(os.Stdout)
+		common.PrintCmdUsage(cmd)
+		return errors.New("please check output to see error")
+	}
+	name := common.GetTaskNameFromArgOrFile(cmd.Flags().Arg(0))
+
+	sources, err := common.GetSourceArgs(cmd)
+	if err != nil {
+		return err
+	}
+
+	resp, err := common.OperateTask(taskOp, name, sources)
+	if err != nil {
+		common.PrintLinesf("can not %s task %s", strings.ToLower(taskOp.String()), name)
+		return err
+	}
+
+	common.PrettyPrintResponse(resp)
+	return nil
+}
+
+func addOperateSourceTaskFlags(cmd *cobra.Command) {
+	cmd.Flags().Bool("all", false, "operate on all (sub)tasks bound to the source")
+	// control workload to dm-cluster for sources with large number of tasks.
+	cmd.Flags().Int(batchSizeFlag, defaultBatchSize, "batch size")
+}
+
+func operateSourceTaskFunc(taskOp pb.TaskOp, cmd *cobra.Command) error {
+	source, batchSize, ok := parseOperateSourceTaskParams(cmd)
+	if !ok {
+		cmd.SetOut(os.Stdout)
+		common.PrintCmdUsage(cmd)
+		return errors.New("please check output to see error")
+	}
+
+	sources := []string{source}
+	ctx, cancel := context.WithTimeout(context.Background(), common.GlobalConfig().RPCTimeout)
+	defer cancel()
+
+	req := pb.QueryStatusListRequest{Sources: sources}
+	resp := &pb.QueryStatusListResponse{}
+	if err := common.SendRequest(ctx, "QueryStatus", &req, &resp); err != nil {
+		common.PrintLinesf("cannot query status of source: %v", sources)
+		return err
+	}
+
+	if !resp.Result || len(resp.Sources) == 0 {
+		common.PrettyPrintInterface(&batchTaskResult{Result: false, Msg: resp.Msg, BatchResults: []*operateTaskResult{}})
+		return nil
+	}
+
+	result := batchOperateTask(taskOp, batchSize, sources, resp.Sources[0].SubTaskStatus)
+	common.PrettyPrintInterface(result)
+
+	return nil
+}
+
+func batchOperateTask(taskOp pb.TaskOp, batchSize int, sources []string, subTaskStatus []*pb.SubTaskStatus) *batchTaskResult {
```

Comment:
maybe move this func to `dm/master/server.go` and make this func to public, so that other package  such as `openapi` could reuse this batch-operate-task logic?

Reply:
or we can implement this `batch-op` in `dm-master` and add a new rpc  instead of calling `dm-master-server` muti times ?

Reply:
`batch-op` is composite task which can be done with multiple `atomic` api calls(`query-status` then `xxx-task`), i think we should not add such an api unless it's very necessary(such as performance bottleneck), or else the number of api would explode